www/content/essays/web-security-basics-with-htmx.md

```diff
@@ -0,0 +1,322 @@
+3. Only serve user-generated content inside HTML tags
+4. If you have authentication cookies, set them with `Secure`, `HttpOnly`, and `SameSite=Lax`
+
+In the following section, I'll discuss what each of these rules does, and what kinds of attack they protect against. The vast majority of htmx users—those using htmx to build a website that allows users to login, view some data, and update that data—should never have any reason to break them.
+
+Later on I will discuss how to break some of these rules. Many useful applications can be built under these constraints, but if you do need more advanced behavior, you'll be doing so with the full knowledge that you're increasing the conceptual burden of securing your application. And you'll have learned a lot about web security in the process.
+
+## Understanding the Rules
+
+### Only call routes you control
+
+This is the most basic one, and the most important: **do not call untrusted routes with htmx.**
+
+In practice, this means you should only use relative URLs. This is fine:
+
+```html
+<button hx-get="/events">Search events</button>
+```
+
+But this is not:
+
+```html
+<button hx-get="https://google.com/search?q=events">Search events</button>
+```
+
+The reason for this is simple: htmx inserts the response from that route directly into the user's page. If the response has a malicious `<script>` inside it, that script can steal the user's data. When you don't control the route, you cannot guarantee that whoever does control the route won't add a malicious script.
+
+Fortunately, this is a very easy rule to follow. Hypermedia APIs (i.e. HTML) are [specific to the layout of your application](https://htmx.org/essays/hypermedia-apis-vs-data-apis/), so there is almost never any reason you'd *want* to insert someone else's HTML into your page. All you have to do is make sure you only call your own routes.
+
+Though it's not quite as popular these days, a common SPA pattern was to separate the frontend and backend into different repositories, and sometimes even to serve them from different URLs. This would require using absolute URLs in the frontend, and often, [disabling CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS). With htmx (and, to be fair, modern React with NextJS) this is an anti-pattern.
+
+Instead, you simply serve your HTML frontend from the same server (or at least the same domain) as your backend, and everything else falls into place: you can use relative URLs, you'll never have trouble with CORS, and you'll never call anyone else's backend.
+
+htmx executes HTML; HTML is code; never execute untrusted code.
+
+### Always use an auto-escaping template engine
+
+When you send HTML to the user, all dynamic content must be escaped. Use a template engine to construct your responses, and make sure that auto-escaping is on.
+
+Fortunately, all template engines support escaping HTML, and most of them enable it by default (Jinja is a notable exception). Below are just a few examples.
```

Comment:
Maybe it should be noted that any web framework that uses Jinja2 will enable auto-escaping for its HTML templates: https://jinja.palletsprojects.com/en/3.0.x/api/#autoescaping

Reply:
This still says it's disabled by default I'm afraid.

Reply:
Well yes, but that’s not the point I was making 🙂 

The jinja2 library does not escape HTML by default, because it’s a generic template system for text files. But any web framework will configure jinja2 to escape in its HTML templates.

Reply:
Oh, I see your point! Thanks for clarifying. I will update the text to clarify.